torch/multiprocessing/spawn.py

```diff
@@ -121,24 +130,67 @@ def join(self, timeout=None):
                 process.terminate()
             process.join()
 
+        def _get_status(b: int) -> str:
+            if b == 0x00:
+                return "Not stuck"
```

Comment:
I don't think we should have all of these changes in torch multiprocessing which are very specific to ProcessGroupNCCL. torch.multiprocessing is a generic package that can be used for a variety of different use cases where NCCL isn't used at all.

All of our logic should probably be encapsulated in ProcessGroupNCCL itself.

Reply:
To actually process the files at the end, I think we can enhance `torch/distributed/launcher` to process these files at the end and determine exactly why the job is stuck.

Although, many times a job just gets stuck and doesn't crash. So it would be useful to have a script that takes the rank_status file as input and processes it to determine which rank is stuck.

Reply:
Do other backends like gloo have some notion of rank statuses? Would it be reasonable to make the other backends always return "Not Set" for all rank statuses for now?

Where in `torch/distributed/launcher` would we add this logic?

> many times a job just gets stuck and doesn't crash.

It would eventually be killed after a timeout, is that correct?

Reply:
+1 for not having rank status specific code in torch multiprocessing code, torch multiprocessing can even be used when there is no distributed APIs being used, this code might make it quite confusing to readers/maintainers. 

For OSS it may not be the case that jobs will be killed, but for internal I think that should be the case afaik as they would be taken down with NCCL_ASYNC_ERROR_HANDLING.